rest/config.go

```diff
@@ -827,13 +828,75 @@ func setupServerContext(config *StartupConfig, persistentConfig bool) (*ServerCo
 		base.Infof(base.KeyAll, "successfully connected to cluster")
 		sc.bootstrapConnection = cluster
 
-		// TODO: CBG-1457 Synchronously find configs in buckets
-		// sc.fetchConfigs()
+		if err := sc.fetchConfigs(); err != nil {
+			return nil, err
+		}
 	}
 
 	return sc, nil
 }
 
+func (sc *ServerContext) fetchConfigs() error {
+	locations, err := sc.bootstrapConnection.GetConfigLocations()
+	if err != nil {
+		return fmt.Errorf("couldn't get buckets from cluster: %w", err)
+	}
+
+	for _, location := range locations {
+		base.Tracef(base.KeyAll, "Checking %q for Sync Gateway config in group %q", location, sc.config.Bootstrap.ConfigGroupID)
+		// populate values from bootstrap, before we overwrite them with any values fetched from the bucket.
+		cnf := DbConfig{
+			BucketConfig: BucketConfig{
+				Server:     &sc.config.Bootstrap.Server,
+				Bucket:     &location,
+				Username:   sc.config.Bootstrap.Username,
+				Password:   sc.config.Bootstrap.Password,
+				CertPath:   sc.config.Bootstrap.X509CertPath,
+				KeyPath:    sc.config.Bootstrap.X509KeyPath,
+				CACertPath: sc.config.Bootstrap.CACertPath,
+			},
+		}
+
+		cas, err := sc.bootstrapConnection.GetConfig(location, sc.config.Bootstrap.ConfigGroupID, &cnf)
+		if errors.As(err, &gocb.ErrDocumentNotFound) {
+			base.Debugf(base.KeyAll, "%q did not contain config in group %q", location, sc.config.Bootstrap.ConfigGroupID)
+			continue
+		}
+		if err != nil {
+			return fmt.Errorf("couldn't fetch config in group %q from location %q: %w", sc.config.Bootstrap.ConfigGroupID, location, err)
+		}
+
+		base.Tracef(base.KeyAll, "Got config for location %q with cas %d", location, cas)
+
+		sc.lock.Lock()
```

Comment:
Given that there are a number of exit paths, maybe paying for a defer Unlock here is worth it?

Reply:
It was because I was locking inside the loop to apply fetched configs.

I've updated so that configs are fetched from the bucket without locking, and then do a 2nd loop which acquires the lock once to apply all the configs.